File: devel/markov_bench.cpp
/*************************************************
 * markov_bench.cpp
 *
 * A C++ program demonstrating:
 *   1. Generating int16_t data (size N).
 *   2. Building a design matrix in float32.
 *   3. Solving for coefficients in float32 with Eigen.
 *   4. Predicting a new sequence (float32 -> round to int16_t).
 *   5. Measuring times & throughput (MB/s).
 *
 * Requires Eigen for linear algebra:
 *   https://eigen.tuxfamily.org/
 *************************************************/

#include <Eigen/Dense>
#include <iostream>
#include <vector>
#include <random>
#include <chrono>
#include <cmath>
#include <cstdint>

//---------------------------------------------------------------------------
// 1) Helper for timing
//---------------------------------------------------------------------------
inline double secondsBetween(
    const std::chrono::high_resolution_clock::time_point& start,
    const std::chrono::high_resolution_clock::time_point& end)
{
    return std::chrono::duration<double>(end - start).count();
}

//---------------------------------------------------------------------------
// 2) Generate random int16_t data
//    We create N random integers in [-30000, 30000].
//---------------------------------------------------------------------------
std::vector<int16_t> generateData(std::size_t N, unsigned seed = 0)
{
    std::mt19937_64 rng(seed);
    std::uniform_int_distribution<int> dist(-30000, 30000);

    std::vector<int16_t> x(N);
    for (std::size_t i = 0; i < N; ++i) {
        x[i] = static_cast<int16_t>(dist(rng));
    }
    return x;
}

//---------------------------------------------------------------------------
// 3) Build the design matrix A (float32) in a naive manner.
//
//    For a Markov model of order M, we want rows for j in [M..N-1]:
//       [1, x_{j-1}, x_{j-2}, ..., x_{j-M} ]
//    We'll store these in A, shape: (N - M) x (M + 1).
//---------------------------------------------------------------------------
Eigen::MatrixXf buildDesignMatrixNaive(const std::vector<int16_t>& x, std::size_t M)
{
    const std::size_t N = x.size();
    const std::size_t rows = N - M;
    const std::size_t cols = M + 1;

    Eigen::MatrixXf A(rows, cols);  // float32

    // Fill A
    for (std::size_t row = 0; row < rows; ++row) {
        // First column = 1.0
        A(row, 0) = 1.0f;
        // Next columns: x_{(row + M) - k}, for k=1..M
        for (std::size_t k = 1; k <= M; ++k) {
            int16_t val = x[(row + M) - k];
            A(row, k) = static_cast<float>(val); // int16 -> float32
        }
    }

    return A;
}

//---------------------------------------------------------------------------
// 4) Solve for coefficients c in float32 using Eigen's SVD.
//
//    c = argmin_c ||A c - y||^2
//    We store and return c as Eigen::VectorXf (float32).
//---------------------------------------------------------------------------
Eigen::VectorXf solveCoeffsEigen(const Eigen::MatrixXf& A, const Eigen::VectorXf& y)
{
    // SVD-based solve (ComputeThinU|V for full solution in least-squares sense)
    // Alternatively: A.colPivHouseholderQr().solve(y), etc.
    Eigen::VectorXf c = A.bdcSvd(Eigen::ComputeThinU | Eigen::ComputeThinV).solve(y);
    return c;
}

//---------------------------------------------------------------------------
// 5) Predict a new sequence using naive float32 arithmetic and round to int16_t.
//
//    - coeffs: c0..cM  (length M+1)
//    - seed:   last M real data points (int16_t)
//    - Output: N predicted points, as int16_t.
//
//    predicted[j] = c0 + c1*predicted[j-1] + ... + cM*predicted[j-M]
//---------------------------------------------------------------------------
std::vector<int16_t> predictNaive(
    const Eigen::VectorXf& coeffs,
    const std::vector<int16_t>& seed,
    std::size_t N)
{
    std::size_t M = coeffs.size() - 1; // since c has M+1 entries
    std::vector<float> predFloat(N, 0.0f);  // store intermediate in float32
    std::vector<int16_t> predInt(N, 0);     // final integer output

    // Initialize first M from seed (cast to float)
    for (std::size_t i = 0; i < M; ++i) {
        predFloat[i] = static_cast<float>(seed[i]);
    }

    // Predict forward
    for (std::size_t j = M; j < N; ++j) {
        float val = coeffs(0); // c0
        for (std::size_t k = 1; k <= M; ++k) {
            val += coeffs(k) * predFloat[j - k];
        }
        predFloat[j] = val;
    }

    // Round to int16_t
    for (std::size_t i = 0; i < N; ++i) {
        float r = std::round(predFloat[i]);
        // clamp into int16 range if you want to be safe, but ignoring extremes here:
        if (r > 32767.f) r = 32767.f;
        if (r < -32768.f) r = -32768.f;
        predInt[i] = static_cast<int16_t>(r);
    }

    return predInt;
}

//---------------------------------------------------------------------------
// 6) Main benchmark function
//---------------------------------------------------------------------------
int main()
{
    // Parameters
    std::size_t N = 5*1000*1000;
    std::size_t M = 5;
    unsigned seed = 42;

    // Generate data
    auto startAll = std::chrono::high_resolution_clock::now();
    auto x = generateData(N, seed);
    auto endAll = std::chrono::high_resolution_clock::now();
    double dtGen = secondsBetween(startAll, endAll);

    // We'll define "bytes processed" as 2*N for throughput (since int16_t=2 bytes).
    double bytesProcessed = double(2 * N);

    std::cout << "\n=== C++ Markov Model Benchmark ===\n"
              << "N=" << N << ", M=" << M << ", data=int16_t, coeffs=float32\n\n";

    std::cout << "Data generation: " << dtGen << " s\n";

    //----------------------------------------------------------------------
    // Build design matrix (naive)
    //----------------------------------------------------------------------
    auto t0 = std::chrono::high_resolution_clock::now();
    Eigen::MatrixXf A = buildDesignMatrixNaive(x, M);
    auto t1 = std::chrono::high_resolution_clock::now();
    double dtBuild = secondsBetween(t0, t1);

    // Prepare y (float32) = x[M..N-1]
    // We'll store it in an Eigen vector
    std::size_t rows = N - M;
    Eigen::VectorXf y(rows);
    for (std::size_t i = 0; i < rows; ++i) {
        y(i) = static_cast<float>(x[i + M]);
    }

    //----------------------------------------------------------------------
    // Solve for coefficients
    //----------------------------------------------------------------------
    auto t2 = std::chrono::high_resolution_clock::now();
    Eigen::VectorXf coeffs = solveCoeffsEigen(A, y);
    auto t3 = std::chrono::high_resolution_clock::now();
    double dtSolve = secondsBetween(t2, t3);

    double dtFittingTotal = dtBuild + dtSolve;

    double buildThroughput = bytesProcessed / (dtBuild * 1.0e6);
    double solveThroughput = bytesProcessed / (dtSolve * 1.0e6);
    double totalThroughput = bytesProcessed / (dtFittingTotal * 1.0e6);

    std::cout << "--- Fitting (Naive build + Eigen solve) ---\n";
    std::cout << "  Build matrix: " << dtBuild << " s, ~" << buildThroughput << " MB/s\n";
    std::cout << "  Solve:        " << dtSolve << " s, ~" << solveThroughput << " MB/s\n";
    std::cout << "  TOTAL:        " << dtFittingTotal << " s, ~" << totalThroughput << " MB/s\n";

    //----------------------------------------------------------------------
    // Prediction
    //----------------------------------------------------------------------
    // We'll seed from the last M points of x
    std::vector<int16_t> seedVals(M);
    for (std::size_t i = 0; i < M; ++i) {
        seedVals[i] = x[N - M + i];
    }

    auto t4 = std::chrono::high_resolution_clock::now();
    auto predictions = predictNaive(coeffs, seedVals, N);
    auto t5 = std::chrono::high_resolution_clock::now();
    double dtPredict = secondsBetween(t4, t5);
    double predThroughput = bytesProcessed / (dtPredict * 1.0e6);

    std::cout << "\n--- Prediction (Naive float32 -> round int16) ---\n";
    std::cout << "  Predict: " << dtPredict << " s, ~" << predThroughput << " MB/s\n";

    //----------------------------------------------------------------------
    // Done
    //----------------------------------------------------------------------
    std::cout << "\n=== Done ===\n\n";

    return 0;
}

File: devel/markov_bench.py
import numpy as np
import time
from sklearn.linear_model import LinearRegression

########################################################################
# 1) Functions to build the design matrix in float32
########################################################################

def build_design_matrix_naive(x: np.ndarray, M: int):
    """
    Naive Python approach to build design matrix A for Markov model.
    x is int16, but we'll cast to float32 in A.
    A has shape (N - M, M + 1).
    Row j = [1, x_{j+M-1}, x_{j+M-2}, ..., x_{j+M-M}],
    for j in 0..(N-M-1).
    """
    N = len(x)
    A = np.ones((N - M, M + 1), dtype=np.float32)  # store as float32
    for row_idx in range(N - M):
        for k in range(1, M + 1):
            # cast x (int16) to float32 on assignment
            A[row_idx, k] = x[M + row_idx - k]
    return A

def build_design_matrix_vectorized(x: np.ndarray, M: int):
    """
    Vectorized approach (using slicing) to build the same design matrix in float32.
    x is int16; cast slices to float32.
    """
    N = len(x)
    rows = N - M
    # First column of 1's in float32
    A_cols = [np.ones(rows, dtype=np.float32)]
    # Next columns are shifted versions of x, cast to float32
    for shift in range(1, M + 1):
        A_cols.append(x[M - shift : N - shift].astype(np.float32))
    A = np.column_stack(A_cols)
    return A

########################################################################
# 2) Functions to solve for coefficients (stored as float32)
########################################################################

def solve_lstsq_numpy(A: np.ndarray, y: np.ndarray):
    """
    Solve for coefficients using NumPy's lstsq solver in float32.
    Returns a 1D array of shape (M+1,) in float32.
    """
    # lstsq may do computations in float64 internally, but we can cast back
    c, _, _, _ = np.linalg.lstsq(A, y, rcond=None)
    return c.astype(np.float32)

def solve_lstsq_sklearn(A: np.ndarray, y: np.ndarray):
    """
    Solve for coefficients using scikit-learn's LinearRegression.
    (fit_intercept=False because we already have the 1-column in A).
    We store final coefficients as float32.
    """
    reg = LinearRegression(fit_intercept=False)
    reg.fit(A, y)
    # sklearn might do double precision internally; cast to float32
    return reg.coef_.astype(np.float32)

########################################################################
# 3) Functions to predict a new sequence
########################################################################

def predict_naive(coeffs: np.ndarray, seed: np.ndarray, N: int):
    """
    Naive Python prediction loop with float32 arithmetic.
    - coeffs: float32 array of shape (M+1,) => [c0, c1, ..., cM].
    - seed: last M known values (int16), but we'll cast to float32 internally.
    - N: number of output points to produce.
    The final predictions are rounded to int16.
    """
    M = len(coeffs) - 1
    # We'll store predictions in float32, then round to int16 at the end
    pred = np.zeros(N, dtype=np.float32)
    # Copy seed (cast to float32)
    pred[:M] = seed.astype(np.float32)

    # Predict forward
    for j in range(M, N):
        val = coeffs[0]
        for k in range(1, M + 1):
            val += coeffs[k] * pred[j - k]
        pred[j] = val

    # Now round to int16
    return np.round(pred).astype(np.int16)

# Optional: Numba-accelerated version
try:
    from numba import njit

    @njit
    def _predict_markov_numba(pred, coeffs, M):
        for j in range(M, len(pred)):
            val = coeffs[0]
            for k in range(1, M + 1):
                val += coeffs[k] * pred[j - k]
            pred[j] = val

    def predict_numba(coeffs: np.ndarray, seed: np.ndarray, N: int):
        """
        Numba-accelerated version of predict_naive (float32 internal arithmetic).
        Rounds final results to int16.
        """
        M = len(coeffs) - 1
        pred = np.zeros(N, dtype=np.float32)
        pred[:M] = seed.astype(np.float32)
        _predict_markov_numba(pred, coeffs, M)
        return np.round(pred).astype(np.int16)

except ImportError:
    def predict_numba(coeffs: np.ndarray, seed: np.ndarray, N: int):
        """
        Fallback if Numba is not installed: use naive approach.
        """
        print("Numba not installed. Falling back to naive prediction.")
        return predict_naive(coeffs, seed, N)

########################################################################
# 4) Main benchmarking function
########################################################################

def benchmark_markov_model(N=1_000_000, M=5, seed=0):
    """
    1) Generate random data x of length N as int16.
    2) Fit the Markov model in multiple ways:
       - Naive (build + solve) in float32
       - Vectorized+NumPy (float32)
       - Vectorized+sklearn (float32)
       Print total fitting time for each method, plus throughput in MB/s
       (using 2*N bytes as the data size).
    3) Predict a new sequence in multiple ways (naive vs numba-accelerated),
       using float32 arithmetic internally, then rounding to int16.
       Print total prediction time + throughput in MB/s.
    """
    np.random.seed(seed)
    # Generate random integers in range [-32768, 32767], but let's just do typical range
    x = np.random.randint(-30000, 30000, size=N, dtype=np.int16)

    # We define "bytes processed" as 2*N because each x_i is an int16 (2 bytes).
    bytes_processed = 2 * N
    # Build target array (still in float32 for the solver)
    # y corresponds to x[M:], but we cast to float32
    y = x[M:].astype(np.float32)

    print(f"\n=== Benchmarking Markov Model (N={N}, M={M}, data=int16, coeffs=float32) ===")

    ########################################################################
    # Fitting: Naive approach
    ########################################################################
    fit_start = time.perf_counter()

    # a) build design matrix (naive)
    t0 = time.perf_counter()
    A_naive = build_design_matrix_naive(x, M)
    dt_build_naive = time.perf_counter() - t0

    # b) solve
    t0 = time.perf_counter()
    c_naive = solve_lstsq_numpy(A_naive, y)
    dt_solve_naive = time.perf_counter() - t0

    total_naive = time.perf_counter() - fit_start

    throughput_build_naive = bytes_processed / (dt_build_naive * 1e6)
    throughput_solve_naive = bytes_processed / (dt_solve_naive * 1e6)
    throughput_total_naive = bytes_processed / (total_naive * 1e6)

    print("\n--- Fitting (Naive) ---")
    print(f"  Build matrix: {dt_build_naive:.3f}s, ~{throughput_build_naive:.2f} MB/s")
    print(f"  Solve:        {dt_solve_naive:.3f}s, ~{throughput_solve_naive:.2f} MB/s")
    print(f"  TOTAL:        {total_naive:.3f}s, ~{throughput_total_naive:.2f} MB/s")

    ########################################################################
    # Fitting: Vectorized + NumPy
    ########################################################################
    fit_start = time.perf_counter()

    # a) build design matrix (vectorized)
    t0 = time.perf_counter()
    A_vec = build_design_matrix_vectorized(x, M)
    dt_build_vec = time.perf_counter() - t0

    # b) solve
    t0 = time.perf_counter()
    c_vec = solve_lstsq_numpy(A_vec, y)
    dt_solve_vec = time.perf_counter() - t0

    total_vec = time.perf_counter() - fit_start

    throughput_build_vec = bytes_processed / (dt_build_vec * 1e6)
    throughput_solve_vec = bytes_processed / (dt_solve_vec * 1e6)
    throughput_total_vec = bytes_processed / (total_vec * 1e6)

    print("\n--- Fitting (Vectorized + NumPy) ---")
    print(f"  Build matrix: {dt_build_vec:.3f}s, ~{throughput_build_vec:.2f} MB/s")
    print(f"  Solve:        {dt_solve_vec:.3f}s, ~{throughput_solve_vec:.2f} MB/s")
    print(f"  TOTAL:        {total_vec:.3f}s, ~{throughput_total_vec:.2f} MB/s")

    ########################################################################
    # Fitting: Vectorized + scikit-learn
    ########################################################################
    fit_start = time.perf_counter()

    # We can re-use A_vec from above, but to measure a true "total time"
    # for the vectorized+sklearn path, let's rebuild it.
    t0 = time.perf_counter()
    A_vec_sklearn = build_design_matrix_vectorized(x, M)
    dt_build_vec_sklearn = time.perf_counter() - t0

    t0 = time.perf_counter()
    c_sklearn = solve_lstsq_sklearn(A_vec_sklearn, y)
    dt_solve_sklearn = time.perf_counter() - t0

    total_sklearn = time.perf_counter() - fit_start

    throughput_build_sklearn = bytes_processed / (dt_build_vec_sklearn * 1e6)
    throughput_solve_sklearn = bytes_processed / (dt_solve_sklearn * 1e6)
    throughput_total_sklearn = bytes_processed / (total_sklearn * 1e6)

    print("\n--- Fitting (Vectorized + Sklearn) ---")
    print(f"  Build matrix: {dt_build_vec_sklearn:.3f}s, ~{throughput_build_sklearn:.2f} MB/s")
    print(f"  Solve:        {dt_solve_sklearn:.3f}s, ~{throughput_solve_sklearn:.2f} MB/s")
    print(f"  TOTAL:        {total_sklearn:.3f}s, ~{throughput_total_sklearn:.2f} MB/s")

    ########################################################################
    # Prediction
    ########################################################################
    # We can pick c_vec from above as "our" coefficients for prediction.
    # We'll seed from the last M points of x (int16).
    seed_vals = x[-M:]

    # 1) Naive Prediction
    t0 = time.perf_counter()
    pred_naive = predict_naive(c_vec, seed_vals, N)
    dt_pred_naive = time.perf_counter() - t0
    throughput_pred_naive = bytes_processed / (dt_pred_naive * 1e6)

    print("\n--- Prediction (Naive) ---")
    print(f"  Predict: {dt_pred_naive:.3f}s, ~{throughput_pred_naive:.2f} MB/s")

    # Warm up Numba
    _ = predict_numba(c_vec, seed_vals, len(seed_vals) + 1)

    t0 = time.perf_counter()
    pred_numba = predict_numba(c_vec, seed_vals, N)
    dt_pred_numba = time.perf_counter() - t0
    throughput_pred_numba = bytes_processed / (dt_pred_numba * 1e6)

    print("\n--- Prediction (Numba) ---")
    print(f"  Predict: {dt_pred_numba:.3f}s, ~{throughput_pred_numba:.2f} MB/s")

    print("\n=== Done ===\n")

########################################################################
# Script entry point
########################################################################
if __name__ == "__main__":
    # Example usage
    benchmark_markov_model(N=5_000_000, M=5, seed=42)
    # Adjust N, M, and seed as needed.
